File: development/Java/samples/client/DAConsole/src/daConsole/Console.java
package daConsole;

import java.util.*;
import com.softing.opc.client.enums.EnumQuality;
import com.softing.opc.client.enums.EnumResultCode;
import com.softing.opc.client.enums.ResultCode;
import java.io.*;

public class Console {

	private static OpcClient m_opcClient = null;
	public static boolean End = false;
	
	public OpcClient getOpcClient()
	{
		return m_opcClient;
	}
	
	public void CreateOpcClient()
	{
		if (m_opcClient == null)
		{
			m_opcClient = new OpcClient();
		}
	}
	public static void Usage()
	{

		System.out.println();
		System.out.println("Usage:");
		System.out.println("Use lower case letters for synchronous execution ");
		System.out.println("Use upper case letters for asynchronous execution ");
		System.out.println("Press \'a\' or \'A\' to activate the objects");
		System.out.println("Press \'c\' or \'C\' to connect the objects");
		System.out.println("Press \'d\' or \'D\' to disconnect the objects");
		System.out.println("Press \'r\' or \'R\' to read the items");
		System.out.println("Press \'w\' or \'W\' to write the items");
		System.out.println("Press \'z\' or \'Z\' to read using server object");
		System.out.println("Press \'t\' or \'T\'to write using the server object");
		System.out.println("Press \'s\' or \'S\' to get the server status");
		System.out.println("Press \'m\' to activate the connection monitor");
		System.out.println("Press \'n\' to deactivate the connection monitor");			
		System.out.println("Press \'?\' or \"u\" to display this usage information");
		System.out.println("Press \'e\' or \'q\' or Ctrl-C to exit");
		System.out.println();
	}
	
	public static void main(String[] args){
		try{
			
			int result = EnumResultCode.S_OK.toInt();
			Console console = new Console();
			console.CreateOpcClient();
			OpcClient m_opcClient = console.getOpcClient();
            
			//	initialize the client instance
			if (!ResultCode.SUCCEEDED(m_opcClient.initialize()))
			{				
				m_opcClient = null;	
				return;
			}	//	end if
			
			//	initialize the DA client simulation
			result |= m_opcClient.initializeDaObjects();
			
			Usage();
			
			boolean end = false;
			while (!end && !Console.End)
			{
				char myChar = (char) System.in.read();
				switch (myChar)
				{
					case 'A':
						m_opcClient.activateSession(false);											
						break;
					case 'a': 
						m_opcClient.activateSession(true);
						break;
					case 'C':
						m_opcClient.connectSession(false);					
						break;
					case 'c': 
						m_opcClient.connectSession(true);
						break;
					case 'D':
						m_opcClient.disconnectSession(false);				
						break;
					case 'd': 
						m_opcClient.disconnectSession(true);
						break;
					case 'R':
						m_opcClient.readItems(true);
						break;
					case 'r':
						m_opcClient.readItems(false);			
						break;
					case 'W': 
						m_opcClient.writeItems(true);
						break;
					case 'w':
						m_opcClient.writeItems(false);																
						break;
					case 'Z':
						m_opcClient.readUsingSession(true);
						break;							
					case 'z':
						m_opcClient.readUsingSession(false);							
						break;
					case 'T':
						m_opcClient.writeUsingSession(true);
						break;
					case 't':
						m_opcClient.writeUsingSession(false);
						break;
					case 'S':
						m_opcClient.getServerStatus(true);
						break;
					case 's':
						m_opcClient.getServerStatus(false);
						break;
                    case 'M':
					case 'm':
						m_opcClient.activateConnectionMonitor();
						break;
					case 'N':
					case 'n':
						m_opcClient.deactivateConnectionMonitor();					
						break;							
					case 'U':							
						break;

					case 'u':
					case '?':
						Usage();				
						break;

					case 'E':
					case 'e':
					case 'Q':
					case 'q':
						end = true;							
						break;

					default: 
						Usage();							
						break;
				}
			}// while


			m_opcClient.terminate();
			m_opcClient = null;
		}
		catch(Exception exc)
		{				
			System.out.println(exc.getMessage());
			exc.printStackTrace();
		}	//	end try...catch
	}
}

File: development/C++/samples/client/DAConsole/Console.cpp
#include "OSCompat.h"
#include "stdafx.h"

#include "OpcClient.h"

#include <iostream>
#ifdef TBC_OS_WINDOWS
#include <conio.h>
#endif

#include <stdio.h>

#ifdef TBC_OS_LINUX
#include <pthread.h>
#include <semaphore.h>
#include <errno.h>
#endif

using namespace std;
using namespace SoftingOPCToolboxClient;
#ifdef TBC_OS_WINDOWS
HANDLE g_endEvent = 0;
#endif
#ifdef TBC_OS_LINUX
sem_t g_endSemaphor;
#endif

#ifdef TBC_OS_WINDOWS

BOOL WINAPI ControlHandler(IN unsigned long crtlType)   // control type
{
	switch (crtlType)
	{
	case CTRL_BREAK_EVENT:  // use Ctrl+C or Ctrl+Break to end server
	case CTRL_C_EVENT:
	case CTRL_CLOSE_EVENT:
		if (g_endEvent)
		{
			SetEvent(g_endEvent);
		}

		return TRUE;
	}

	return FALSE;
} // end controlHandler
#endif

void usage(void)
{
	_tprintf(_T("\nUsage:\n"));
	_tprintf(_T("Use lower case letters for synchronous execution\n"));
	_tprintf(_T("Use upper case letters for asynchronous execution\n"));
	_tprintf(_T("Press \'a\' or \'A\' to activate the objects\n"));
	_tprintf(_T("Press \'c\' or \'C\' to connect the objects\n"));
	_tprintf(_T("Press \'d\' or \'D\' to disconnect the objects\n"));
	_tprintf(_T("Press \'r\' or \'R\' to read the items\n"));
	_tprintf(_T("Press \'w\' or \'W\' to write the items\n"));
	_tprintf(_T("Press \'F\' to refresh the items\n"));
	_tprintf(_T("Press \'z\' or \'Z\' to read using server object\n"));
	_tprintf(_T("Press \'t\' or \'T\' to write using the server object\n"));
	_tprintf(_T("Press \'s\' or \'S\' to get the server status\n"));
	_tprintf(_T("Press \'h\' to change the subscription`s data retrieval mode\n"));
	_tprintf(_T("Press \'m\' to activate the connection monitor\n"));
	_tprintf(_T("Press \'n\' to deactivate the connection monitor\n"));
	_tprintf(_T("Press \'l\' or \'L\' to logon for secure communication\n"));
	_tprintf(_T("Press \'o\' or \'O\' to logoff for secure communication\n"));
	_tprintf(_T("Press \'?\' or \'u\' to display this usage information\n"));
	_tprintf(_T("Press \'e\' or \'q\' or Ctrl-C to exit\n"));
}

int _tmain(int argc, _TCHAR* argv[])
{
	long result = S_OK;
#ifdef TBC_OS_WINDOWS
	SetConsoleCtrlHandler(ControlHandler, TRUE);
#endif
	// create the OpcClient instance
	createOpcClient();
	OpcClient* pClient = getOpcClient();
#ifdef TBC_OS_WINDOWS
	unsigned long waitRet;
	g_endEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
#endif

	// initialize the client instance
	if (!SUCCEEDED(pClient->initialize()))
	{
		pClient->terminate();
		destroyOpcClient();
#ifdef TBC_OS_WINDOWS
		CloseHandle(g_endEvent);
#endif
		return 1;
	}

	// initialize the DA client simulation
	result |= pClient->initializeDaObjects();
	usage();
	unsigned long waitTime = 100;
#ifdef TBC_OS_WINDOWS
	BOOL end = FALSE;

	while (!end)
	{
		waitRet = WaitForSingleObject(g_endEvent, waitTime);
		waitTime = 100;

		if (waitRet == WAIT_OBJECT_0)
		{
			end = TRUE;
			CloseHandle(g_endEvent);
		}
		else
		{
#endif
#ifdef TBC_OS_LINUX
	unsigned long timeout = 0;

	while (timeout < waitTime)
	{
		timespec delay;
		delay.tv_sec = 0;
		delay.tv_nsec = 100000000;  /* 100 milli sec */
		// Wait for the event be signaled
		int retCode = sem_trywait(&g_endSemaphor); // event semaphore handle

		if (!retCode)
		{
			/* Event is signaled */
			break;

		}
		else
		{
			/* check whether somebody else has the mutex */
			if (retCode == EPERM)
			{
				/* sleep for delay time */
				nanosleep(&delay, NULL);
				timeout++ ;
			}

#endif
#ifdef TBC_OS_WINDOWS

			if (_kbhit())
			{
				int input;
				input = _getch();
#endif
#ifdef TBC_OS_LINUX
			char input;

			if (1)
			{
				std::cin.get(input);
#endif

				switch (input)
				{
				case _T('A'):
					pClient->activateAsyncSubscription();
					break;

				case _T('a'):
					pClient->activateSyncSubscription();
					break;

				case _T('C'):
					pClient->connectAsyncSubscription();
					break;

				case _T('c'):
					pClient->connectSyncSubscription();
					break;

				case _T('D'):
					pClient->disconnectAsyncSubscription();
					break;

				case _T('d'):
					pClient->disconnectSyncSubscription();
					break;

				case _T('R'):
					pClient->readAsyncSubscription();
					break;

				case _T('r'):
					pClient->readSyncItem();
					pClient->readSyncSubscription();
					break;

				case _T('F'):
				case _T('f'):
					//the refresh can be done only asynchronously!
					pClient->refreshSubscription();
					break;

				case _T('Z'):
					pClient->readAsyncSession();
					break;

				case _T('z'):
					pClient->readSyncSession();
					break;

				case _T('W'):
					pClient->writeAsyncSubscription();
					break;

				case _T('w'):
					pClient->writeSyncSubscription();
					break;

				case _T('T'):
					pClient->writeAsyncSession();
					break;

				case _T('t'):
					pClient->writeSyncSession();
					break;

				case _T('S'):
					pClient->getAsyncServerStatus();
					break;

				case _T('s'):
					pClient->getSyncServerStatus();
					break;

				case _T('h'):
				case _T('H'):
					pClient->changeDataRetrievalMode();
					break;

				case _T('M'):
				case _T('m'):
					pClient->activateConnectionMonitor();
					break;

				case _T('N'):
				case _T('n'):
					pClient->deactivateConnectionMonitor();
					break;

				case _T('l'):
					pClient->setCredentials();
					pClient->logOnSyncForSecureCommunication();
					break;

				case _T('L'):
					pClient->setCredentials();
					pClient->logOnAsyncForSecureCommunication();
					break;

				case _T('o'):
					pClient->setCredentials(_T("Guest"), _T("nopass"));
					pClient->logOffSyncForSecureCommunication();
					break;

				case _T('O'):
					pClient->logOffAsyncForSecureCommunication();
					pClient->setCredentials(_T("Guest"), _T("nopass"));
					break;

				case _T('?'):
				case _T('U'):
				case _T('u'):
					usage();
					break;

				case _T('E'):
				case _T('e'):
				case _T('Q'):
				case _T('q'):
#ifdef TBC_OS_WINDOWS
					end = TRUE;
					CloseHandle(g_endEvent);
#endif
#ifdef TBC_OS_LINUX
					sem_post(&g_endSemaphor);
#endif
					break;
				} // end switch
			} // end if
		} // end else

#ifdef TBC_OS_WINDOWS
		Sleep(10);
#endif
	} // end while

	pClient->terminate();
	destroyOpcClient();
	return 0;
} // end main
